File: apps/web/src/lib/milestones.ts
// Browser-side client for the milestone API (TASK-110).
//
// Types mirror packages/contracts/openapi.json (ADR-008). Requests
// go to /api/* on this origin and are proxied to the backend by
// next.config.ts rewrites.

/** MilestoneResponse from the OpenAPI contract. */
export type MilestoneView = {
  milestone_id: string;
  run_id: string;
  goal_id: string;
  title: string;
  description: string;
  target_date: string;
  created_at: string;
  updated_at: string;
};

export type MilestoneDraft = {
  title: string;
  target_date: string;
  description?: string;
};

async function request<T>(path: string, init?: RequestInit): Promise<T> {
  const response = await fetch(path, {
    headers: { "Content-Type": "application/json" },
    ...init,
  });
  if (!response.ok) {
    let detail = `${response.status} ${response.statusText}`;
    try {
      const body = (await response.json()) as { detail?: unknown };
      if (typeof body.detail === "string") {
        detail = body.detail;
      }
    } catch {
      // not JSON — keep the status line
    }
    throw new Error(detail);
  }
  return (await response.json()) as T;
}

/** Pin a checkpoint on the goal's current backcasting run. */
export function defineMilestone(
  goalId: string,
  draft: MilestoneDraft,
): Promise<MilestoneView> {
  return request<MilestoneView>(`/api/goals/${goalId}/milestones`, {
    method: "POST",
    body: JSON.stringify(draft),
  });
}

File: apps/api/src/backcasting/api/milestones.py
"""Milestone HTTP API (TASK-110).

Exposes the milestone use cases (``backcasting.application.milestones``)
over HTTP, scoped under the goal whose run the checkpoints attach
to.

Status mapping:

- unknown goal → 404
- no backcasting run yet → 404 (define the backcast first)
- domain violations (blank title, target at or before now, a
  finished run) → 422
"""

from __future__ import annotations

import uuid

from fastapi import APIRouter, HTTPException, Request, status
from pydantic import AwareDatetime, BaseModel

from backcasting.application.goals import GoalNotFoundError
from backcasting.application.milestones import (
    MilestoneService,
    NoBackcastRunError,
)
from backcasting.domain.milestone import Milestone, MilestoneError

router = APIRouter(tags=["milestones"])


def _service(request: Request) -> MilestoneService:
    return request.app.state.milestone_service


class MilestoneCreate(BaseModel):
    title: str
    target_date: AwareDatetime
    description: str = ""


class MilestoneResponse(BaseModel):
    milestone_id: uuid.UUID
    run_id: uuid.UUID
    goal_id: uuid.UUID
    title: str
    description: str
    target_date: str
    created_at: str
    updated_at: str


def _milestone_response(milestone: Milestone) -> MilestoneResponse:
    return MilestoneResponse(
        milestone_id=milestone.milestone_id,
        run_id=milestone.run_id,
        goal_id=milestone.goal_id,
        title=milestone.title,
        description=milestone.description,
        target_date=milestone.target_date.isoformat(),
        created_at=milestone.created_at.isoformat(),
        updated_at=milestone.updated_at.isoformat(),
    )


@router.post(
    "/goals/{goal_id}/milestones",
    status_code=status.HTTP_201_CREATED,
    operation_id="defineMilestone",
    response_model=MilestoneResponse,
)
def define_milestone(
    goal_id: uuid.UUID, payload: MilestoneCreate, request: Request
) -> MilestoneResponse:
    """Pin a checkpoint on the goal's current backcasting run
    (docs/04 step 10)."""
    try:
        milestone = _service(request).define_milestone(
            goal_id=goal_id,
            title=payload.title,
            target_date=payload.target_date,
            description=payload.description,
        )
    except (GoalNotFoundError, NoBackcastRunError) as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)
        ) from exc
    except MilestoneError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc
    return _milestone_response(milestone)


@router.get(
    "/goals/{goal_id}/milestones",
    operation_id="listMilestones",
    response_model=list[MilestoneResponse],
)
def list_milestones(
    goal_id: uuid.UUID, request: Request
) -> list[MilestoneResponse]:
    """The goal's checkpoints on its current run, earliest target
    first. Empty while no run exists."""
    milestones = _service(request).list_for_goal(goal_id)
    return [_milestone_response(milestone) for milestone in milestones]
